xpart/particles/particles.py

```diff
@@ -157,109 +159,346 @@ class Particles(xo.HybridClass):
-                pass
+    def _sqrt(self, x):
+        if isinstance(self._context, xo.ContextPyopencl):
+            import pyopencl.clmath as clm
+            return clm.sqrt(x)
+        else:
+            return np.sqrt(x)
+
+    def _assert_values_consistent(self, given_value, computed_value, mask=None):
+        """Check if the given value is consistent with the computed value."""
+        if given_value is None:
+            return
+        if not self._allclose(given_value, computed_value, mask=mask):
+            raise ValueError(
+                f'The given value {given_value} is not consistent with the '
+                f'computed value {computed_value}. Difference: '
+                f'{abs(given_value - computed_value)}.'
+            )
+
+    def _setattr_if_consistent(self, varname, given_value, computed_value,
+                               mask=None):
+        """Update field values that may be both given and computed from others.
+
+        In case of small differences between the given value and them computed
+        value, the given value will prevail to preserve numerical stability.
+        This is useful when two or more dependent variables are given as input.
+        """
+        self._assert_values_consistent(given_value, computed_value, mask)
+        target_val = given_value if given_value is not None else computed_value
+
+        # The simple case
+        if mask is None:
+            setattr(self, varname, target_val)
+            return
+
+        # Assign with a mask
+        if isinstance(self._context, xo.ContextPyopencl):  # PyOpenCL array
+            if hasattr(mask, 'get'):
+                mask = mask.get()
+            mask = np.where(mask)[0]
+            mask = self._context.nparray_to_context_array(mask)
+
+        getattr(self, varname)[mask] = target_val[mask]
+
+    def _update_refs(self, p0c=None, energy0=None, gamma0=None, beta0=None,
+                     mask=None):
+        if not any(ff is not None for ff in (p0c, energy0, gamma0, beta0)):
+            self._p0c = 1e9
+            p0c = self._p0c
+
+        if p0c is not None:
+            _energy0 = self._sqrt(p0c ** 2 + self.mass0 ** 2)
+            _beta0 = p0c / _energy0
+            _gamma0 = _energy0 / self.mass0
+            _p0c = p0c
+        elif energy0 is not None:
+            _p0c = self._sqrt(energy0 ** 2 - self.mass0 ** 2)
+            _beta0 = _p0c / energy0
+            _gamma0 = energy0 / self.mass0
+        elif gamma0 is not None:
+            _beta0 = self._sqrt(1 - 1 / gamma0 ** 2)
+            _energy0 = self.mass0 * gamma0
+            _p0c = _energy0 * _beta0
+            _gamma0 = gamma0
+        elif beta0 is not None:
+            _gamma0 = 1 / self._sqrt(1 - beta0 ** 2)
+            _energy0 = self.mass0 * _gamma0
+            _p0c = _energy0 * beta0
+            _beta0 = beta0
+        else:
+            raise RuntimeError('This statement is unreachable.')
```

Comment:
Check consistency with the physics guide:
https://github.com/xsuite/xsuite/blob/main/docs/physics_manual/physics_man.pdf